File: tests/org.eclipse.rap.rwt.test/src/org/eclipse/swt/widgets/Sash_Test.java
/*******************************************************************************
 * Copyright (c) 2011, 2015 EclipseSource and others.
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v1.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-v10.html
 *
 * Contributors:
 *    EclipseSource - initial API and implementation
 ******************************************************************************/
package org.eclipse.swt.widgets;

import static org.eclipse.rap.rwt.testfixture.internal.SerializationTestUtil.serializeAndDeserialize;
import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertSame;
import static org.junit.Assert.assertTrue;
import static org.mockito.Mockito.mock;

import org.eclipse.rap.rwt.internal.lifecycle.WidgetLCA;
import org.eclipse.rap.rwt.testfixture.TestContext;
import org.eclipse.swt.SWT;
import org.eclipse.swt.events.SelectionListener;
import org.eclipse.swt.internal.widgets.sashkit.SashLCA;
import org.junit.Before;
import org.junit.Rule;
import org.junit.Test;


public class Sash_Test {

  @Rule
  public TestContext context = new TestContext();

  private Composite shell;
  private Sash sash;

  @Before
  public void setUp() {
    Display display = new Display();
    shell = new Shell( display, SWT.NONE );
    sash = new Sash( shell, SWT.NONE );
  }

  @Test
  public void testIsSerializable() throws Exception {
    sash.setSize( 1, 2 );

    Sash deserializedSash = serializeAndDeserialize( sash );

    assertEquals( sash.getSize(), deserializedSash.getSize() );
  }

  @Test
  public void testAddSelectionListener() {
    sash.addSelectionListener( mock( SelectionListener.class ) );

    assertTrue( sash.isListening( SWT.Selection ) );
    assertTrue( sash.isListening( SWT.DefaultSelection ) );
  }

  @Test( expected = IllegalArgumentException.class )
  public void testAddSelectionListener_withNullArgument() {
    sash.addSelectionListener( null );
  }

  @Test
  public void testRemoveSelectionListener() {
    SelectionListener listener = mock( SelectionListener.class );
    sash.addSelectionListener( listener );

    sash.removeSelectionListener( listener );

    assertFalse( sash.isListening( SWT.Selection ) );
    assertFalse( sash.isListening( SWT.DefaultSelection ) );
  }

  @Test( expected = IllegalArgumentException.class )
  public void testRemoveSelectionListener_withNullArgument() {
    sash.removeSelectionListener( null );
  }

  @Test
  public void testGetAdapter_LCA() {
    assertTrue( sash.getAdapter( WidgetLCA.class ) instanceof SashLCA );
    assertSame( sash.getAdapter( WidgetLCA.class ), sash.getAdapter( WidgetLCA.class ) );
  }

}

File: tests/org.eclipse.rap.rwt.jstest/js/org/eclipse/rwt/test/tests/SashTest.js
/*******************************************************************************
 * Copyright (c) 2011, 2014 EclipseSource and others.
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v1.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-v10.html
 *
 * Contributors:
 *    EclipseSource - initial API and implementation
 ******************************************************************************/

(function() {

var TestUtil = org.eclipse.rwt.test.fixture.TestUtil;
var ObjectRegistry = rwt.remote.ObjectRegistry;
var Processor = rwt.remote.MessageProcessor;

rwt.qx.Class.define( "org.eclipse.rwt.test.tests.SashTest", {

  extend : rwt.qx.Object,

  members : {

    testCreateSashByProtocol : function() {
      var shell = TestUtil.createShellByProtocol( "w2" );
      Processor.processOperation( {
        "target" : "w3",
        "action" : "create",
        "type" : "rwt.widgets.Sash",
        "properties" : {
          "style" : [],
          "parent" : "w2"
        }
      } );
      var widget = ObjectRegistry.getObject( "w3" );
      assertTrue( widget instanceof rwt.widgets.Sash );
      assertIdentical( shell, widget.getParent() );
      assertTrue( widget.getUserData( "isControl") );
      assertEquals( "sash", widget.getAppearance() );
      assertEquals( "vertical", widget.getOrientation() );
      shell.destroy();
      widget.destroy();
    },

    testCreateSashHorizontalByProtocol : function() {
      var shell = TestUtil.createShellByProtocol( "w2" );
      Processor.processOperation( {
        "target" : "w3",
        "action" : "create",
        "type" : "rwt.widgets.Sash",
        "properties" : {
          "style" : [ "HORIZONTAL" ],
          "parent" : "w2"
        }
      } );
      var widget = ObjectRegistry.getObject( "w3" );
      assertTrue( widget instanceof rwt.widgets.Sash );
      assertEquals( "horizontal", widget.getOrientation() );
      shell.destroy();
      widget.destroy();
    },

    testSetSelectionListenerByProtocol : function() {
      var shell = TestUtil.createShellByProtocol( "w2" );
      Processor.processOperation( {
        "target" : "w3",
        "action" : "create",
        "type" : "rwt.widgets.Sash",
        "properties" : {
          "style" : [],
          "parent" : "w2"
        }
      } );
      var widget = ObjectRegistry.getObject( "w3" );

      TestUtil.protocolListen( "w3", { "Selection" : true } );

      var remoteObject = rwt.remote.Connection.getInstance().getRemoteObject( widget );
      assertTrue( remoteObject.isListening( "Selection" ) );
      shell.destroy();
      widget.destroy();
    },

    testSendWidgetSelected : function() {
      var shell = TestUtil.createShellByProtocol( "w2" );
      Processor.processOperation( {
        "target" : "w3",
        "action" : "create",
        "type" : "rwt.widgets.Sash",
        "properties" : {
          "style" : [],
          "parent" : "w2",
          "bounds" : [ 10, 20 , 10, 100 ]
        }
      } );
      var widget = ObjectRegistry.getObject( "w3" );
      TestUtil.protocolListen( "w3", { "Selection" : true } );

      this.fakeDrag( widget, 5, 0 );

      var message = TestUtil.getLastMessage();
      assertEquals( 15, message.findNotifyProperty( "w3", "Selection", "x" ) );
      assertEquals( 20, message.findNotifyProperty( "w3", "Selection", "y" ) );
      assertEquals( 10, message.findNotifyProperty( "w3", "Selection", "width" ) );
      assertEquals( 100, message.findNotifyProperty( "w3", "Selection", "height" ) );
      shell.destroy();
    },

    testMouseOver : function() {
      var sash = new rwt.widgets.Sash();
      sash.addToDocument();
      TestUtil.flush();

      TestUtil.mouseOver( sash );
      assertTrue( sash.hasState( "over" ) );

      TestUtil.mouseOut( sash );
      assertFalse( sash.hasState( "over" ) );

      sash.destroy();
    },

    fakeDrag : function( sash, leftOffset, topOffset ) {
      sash._commonMouseDown();
      sash.setLeft( sash.getLeft() + leftOffset );
      sash.setTop( sash.getTop() + topOffset );
      sash._commonMouseUp();
    }

  }

} );

}() );
